src/basic-noise-qvm.lisp

```diff
@@ -0,0 +1,194 @@
+  (declare (ignore args))
+  (maphash #'%check-noise-entry (t1-ops qvm))
+  (maphash #'%check-noise-entry (t2-ops qvm))
+  (maphash #'%check-noise-entry (depol-ops qvm))
+  (maphash #'%check-povm-entry (readout-povms qvm)))
+
+
+(defun %check-noise-entry (qubit noise-ops)
+  "Helper function to check that a key value pair in T1-OPS or T2-OPS is valid. The qubit must be a NON-NEGATIVE INTEGER, and the NOISE-OPS must be a valid kraus map."
+  (assert (and (integerp qubit) (<= 0 qubit)))
+  (check-kraus-ops noise-ops))
+
+
+(defun %check-povm-entry (qubit povm)
+  "Helper fucntion to check that a key value pair in READOUT-POVMS is valid. The QUBIT must be a NON-NEGATIVE INTEGER, and the POVM must be a correctly formatted POVM."
+  (assert (and (integerp qubit) (<= 0 qubit)))
+  (check-povm povm))
+
+
+(defun (setf qubit-t1) (t1 qvm qubit)
+  "Save the kraus operators for the T1 provided on the specific qubit."
+  (let ((kraus-ops (generate-damping-kraus-map t1 (avg-gate-time qvm))))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (t1-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-t2) (t2 qvm qubit)
+  "Save the kraus operators for the T2  provided on the specified qubit."
+  (let ((kraus-ops (generate-dephasing-kraus-map t2 (avg-gate-time qvm))))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (t2-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-depolarization) (depol-prob qvm qubit)
+  "Save the kraus operators for the depolarizing channel on QUBIT. The depolarizing channel represents the average gate overrotation/underrotation noise. It is specified by DEPOL-PROB, which is the probability of overrotation/underrotation for the given QUBIT."
+  (let ((kraus-ops (generate-depolarizing-kraus-map depol-prob)))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (depol-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-fro) (fro qvm qubit)
+  "Save the readout povms for the fro provided on the specified qubit. The fR0 is assumed to be the average readout fidelity (avg of p(1|1) and p(0|0)) so as an approximation, we set p(0|0) = p(1|1) = fR0 in the readout povm assignment probabilities."
+  (setf (gethash qubit (readout-povms qvm)) (list fro (- 1.0d0 fro) (- 1.0d0 fro) fro)))
+
+
+(defmethod apply-all-kraus-maps ((qvm basic-noise-qvm) (instr quil:gate-application) kraus-ops)
+  ;; Apply all the kraus operators for each type of noise
+  ;; source. KRAUS-OPS is a list of ( (t1-operators) (t2-operators)),
+  ;; where t1-operators is a list of t1 noise maps for each qubit in
+  ;; the instruction.  If the instruction is a 2q gate, we need to
+  ;; tensor the kraus operators for each qubit's error source.
+  (loop :for noise-operators :in kraus-ops
+        :do (let ((simplified-kraus-ops  (reduce #'kraus-kron noise-operators)))
+              (when simplified-kraus-ops
+                (apply-kraus-map qvm instr simplified-kraus-ops)))))
+
+
+(defmethod transition ((qvm basic-noise-qvm) (instr quil:gate-application))
```

Comment:
Do you think it would make sense to use an `:after` method combination [1] here? This seems like a conventional transition on which you call `apply-all-kraus-maps` **after** you update the wavefunction. Using an `:after` combination would slightly simplify the code and, more importantly, convey the intention clearly.

[1] http://www.lispworks.com/documentation/HyperSpec/Body/07_ffb.htm

Reply:
I just tried doing this because I agree that it makes sense. However, I ran into issues because `basic-noise-qvm` is a subclass of `channel-qvm`, which is a subclass of `pure-state-qvm`. If I make `basic-noise-qvm`'s transition an `:after` method, it tries to run `channel-qvm`'s transition first, which I do not want. Unless there is a way to prevent this from happening and to default all the way back to `pure-state-qvm`'s transition, then I think I have to keep it the way it is.

Reply:
I see. Even though we could probably create a suitable method combination to accomplish this goal, for the moment I think this can be left as is (perhaps adding a short comment explaining that an `:after` method won't do in this case).

Reply:
I will make an issue for refactoring the `transition` functions for `channel-qvm` and `basic-noise-qvm` to use `:around`/`:after` to avoid repetitive code, but for now I'll leave them as they are